### DISCUSSION TITLE
dynamic connection

### DISCUSSION THREAD

**Discussion by paskuale75** [simialbi/yii2-elfinder#5] [Q&A]
how can i manage the files of different users by adding a parameter to customize the connectionset ? 
thanks

**Comment by simialbi** [comment]
Hi @paskuale75 

Do you have an example what exactly you want to do?

**Reply by paskuale75** [reply]
My application manages a series of containers, I would like to give the possibility to manage files linked to each single container, so the path cannot be unique for everyone, how can I diversify based on an ID ?
thanks

**Reply by simialbi** [reply]
I don't know your configuration, so the code below is more some **pseudo code example**.
You could create a dynamic connection set in an bootstraping component e.g.:
```php
public function bootstrap($app)
{
    $module = $app->getModule('elfinder');
    $myClientModel = Client::find($app->session->user->id);
    $configuration = [
        'class' => '\simialbi\yii2\elfinder\ElFinderConfigurationFTP',
        'host' => $myClientModel->host,
        'user' => $myClientModel->user,
        'pass' => $myClientModel->pass
    ];
    $module->connectionSets['default'] = [$configuration];
}
```

**Reply by paskuale75** [reply]
I don't use ftp, my case is ElFinderConfigurationLocalFileSystem, I would like to create a link similar to the update `Url::to(['// moduleName/controllerName/actionName, 'id => clientID']) `for open a elfinder view for it ? Any example ?
Thanks

**Reply by paskuale75** [reply]
can I create a connectionSet dynamically? Without relying on the bootstrap function ?
thanks

**Reply by simialbi** [reply]
Of course: You can always add one like in the example above. I've you want to add one, instead of overriding all of them just do it like this:
```php
$module->connectionSets['key'][] = new ElFinderConfigurationLocalFileSystem(/*[configuration]*/);
```

**Reply by paskuale75** [reply]
So can i use it dynamically ?
I try with your suggestion:
```
$module = Yii::$app->getModule('elfinder');
$myClientModel = Pratica::find(2);
$configuration = [
    'class' => 'simialbi\yii2\elfinder\ElFinderConfigurationLocalFileSystem',
    'path'  => '@webroot/pratica_files/2',
    'URL'   => '@web/pratica_files/2'
];
//$module->connectionSets['default'] = [$configuration];
$module->connectionSets['myModelWithIdEqualTo31'][] = new ElFinderConfigurationLocalFileSystem([$configuration]);

return me exception:
`Setting unknown property: simialbi\yii2\elfinder\ElFinderConfigurationLocalFileSystem::0`
```

**Reply by simialbi** [reply]
Change
```php
$module->connectionSets['myModelWithIdEqualTo31'][] = new ElFinderConfigurationLocalFileSystem([$configuration]);
```
to
```php
$module->connectionSets['myModelWithIdEqualTo31'][] = new ElFinderConfigurationLocalFileSystem($configuration);
```
(no `[]` inside constructor because `$configuration` already is an array)
and remove `class` property in `$configuration` array. It's not needed here.

**Reply by paskuale75** [reply]
Thanks @simialbi I updated code as you suggest:
```
$module = Yii::$app->getModule('elfinder');
$myClientModel = Pratica::find(2);
$configuration = [
    'path'  => '@webroot/pratica_files/2',
    'URL'   => '@web/pratica_files/2'
];

$module->connectionSets['myModelWithIdEqualTo31'][] = new ElFinderConfigurationLocalFileSystem($configuration);

?>

<!-- Small boxes (Stat box) -->
<div class="row">
    <div class="col-lg-12">
        <?php
        // @see https://github.com/Studio-42/elFinder/wiki/Client-configuration-options-2.1
        echo ElFinder::widget([
            'instanceName' => 'myModelWithIdEqualTo31' // from module connectionSets/volumeBehaviors configuration (array key)
        ]);
        ?>
    </div>....
```

But return me '404 - Page not found' what happens ?

**Reply by simialbi** [reply]
Hmm... Ok you're right. This way the component is not initializing. I'm going to extend the module...

**Reply by paskuale75** [reply]
I'm glad I contributed to giving you an idea to further improve this nice extension :)

**Reply by simialbi** [reply]
I added the [addConnectionSet](https://github.com/simialbi/yii2-elfinder/blob/master/Module.php#L160) method in the module class. Try following (after composer update to [1.7.1](https://github.com/simialbi/yii2-elfinder/releases/tag/1.7.1)):
```php
$module = Yii::$app->getModule('elfinder');
$module->addConnectionSet('myModelWithIdEqualTo31', $configuration);
echo ElFinder::widget([
    'instanceName' => 'myModelWithIdEqualTo31' // from module connectionSets/volumeBehaviors configuration (array key)
]);
```

**Reply by paskuale75** [reply]
thanks for update, but now return me this:
```
Invalid Configuration – yii\base\InvalidConfigException
Object configuration must be an array containing a "class" or "__class" element.
```

my actual code: ( I added class for fix but don't work)

```
$module = Yii::$app->getModule('elfinder');
$myClientModel = Pratica::find(2);
$configuration = [
    'class' => 'simialbi\yii2\elfinder\ElFinderConfigurationLocalFileSystem', 
    'path'  => '@webroot/pratica_files/2',
    'URL'   => '@web/pratica_files/2'
];
$module->addConnectionSet('myModelWithIdEqualTo31', $configuration);
```

**Reply by simialbi** [reply]
I think I know why. Please composer update to 1.7.2 and try again

**Reply by paskuale75** [reply]
> I think I know why. Please composer update to 1.7.2 and try again

with great pleasure :)
this conversation is taking the form of a how-to !

**Reply by simialbi** [reply]
😂

**Reply by paskuale75** [reply]
after 1.7.2 update return this :
`array_key_exists() expects parameter 2 to be array, null given`

**Reply by simialbi** [reply]
Do you have some stack trace?

**Reply by simialbi** [reply]
I think I found it. Should be fixed now (with version `1.7.3`)

**Reply by paskuale75** [reply]
thanks, sorry I had moved away from the pc, new exception return me after 1.7.3 version

`The file or directory to be published does not exist: /var/www/html/yii2-argolab-usuario/vendor/simialbi/yii2-elfinder/assets
`

**Reply by simialbi** [reply]
Structure change generated a new error 🤦‍♂️, fixed with `1.7.4`

**Reply by paskuale75** [reply]
I updated to `1.7.4` now return this (json format)
`{"name":"Not Found","message":"Page not found.","code":0,"status":404,"type":"yii\\web\\NotFoundHttpException"}`

$elfinder return _**null**_  in `vendor/simialbi/yii2-elfinder/src/controllers/ConnectionController.php` file

....and what's more, the layout seems to have suffered some damage.
